### PR TITLE
EXPOSED-372 UpsertStatement.resultedValues contains incorrect value

### PR DESCRIPTION
**Pull request by obabichevjb**
Here is the fix for the issue [EXPOSED-372](https://youtrack.jetbrains.com/issue/EXPOSED-372) 

The problem is that `upsert()` command returns a wrong id in the case of `UUIDTable`. 

The key reason for that difference I found is that for that table the `uuid(columnName).autoGenerate().entityId()` column is used, but inside `InsertStatement` there is a condition `col.columnType.isAutoInc || value is NextVal<*>` for overwriting the value from `arguments`. 

I added `col.defaultValueFun != null` to the condition to avoid hard overwriting of the value from the result set. It fixes the problem from the issue.

But I feel that something wrong happens there, because it's weird to me that we need to rewrite values from result stet at all. I tried to replace the whole block:
```
if (col.columnType.isAutoInc || value is NextVal<*> || col.defaultValueFun != null) {
    map.getOrPut(col) { value }
} else {
    map[col] = value
}
```
with the statement `map.getOrPut(col) { value }` assuming that we take values from arguments only if it's not in the result set.

But after such change, some other tests start to fail (particularly for SQLite), because of the line above:
```
if (returnedValues.isEmpty() && firstAutoIncColumn != null) returnedValues[firstAutoIncColumn] = rs.getObject(1)
```
The value returned from `rs.getObject(1)` is not actual id, but something different. The test `InsertTests::testInsertWithPredefinedId()` fails  on `assertEquals(id1, entityID)` with error `Failed on SQLITE expected:<1> but was:<id1>`. What means that `rs.getObject(1)` returned `1` instead of `id1` (actual value in DB at this moment is `id1`).

### CODE REVIEW

**Comment by obabichevjb** [comment]
It looks like it's the wrong way to fix it, because other tests fail with `Unexpected value of type UUID: 1 of kotlin.Int`. I will try to find another way.

**Comment by obabichevjb** [comment]
It works now but particularly for Postgres. Using upsert command with other DBs will cause the same problem of returning wrong ID. It may confuse users.

File: exposed-tests/src/test/kotlin/org/jetbrains/exposed/sql/tests/shared/dml/UpsertTests.kt [hunk 1]
```diff
@@ -626,6 +627,34 @@ class UpsertTests : DatabaseTestsBase() {
...
+    @Test
+    fun testUpsertWithUUIDPrimaryKey() {
+        val tester = object : UUIDTable("upsert_test", "id") {
+            val key = integer("test_key").uniqueIndex()
+            val value = text("test_value")
+        }
+
+        // At present, only Postgres returns the correct UUID directly from the result set.
+        // For other databases incorrect ID is returned from the 'upsert' command.
+        withTables(excludeSettings = TestDB.entries - listOf(TestDB.POSTGRESQL, TestDB.POSTGRESQLNG), tester) { testDb ->
```
Comment (bog-walk) [inline exposed-tests/src/test/kotlin/org/jetbrains/exposed/sql/tests/shared/dml/UpsertTests.kt:639]: For the future, there are some available sets for common `TestDB` collections in its companion object, if you ever need them.

**Review by bog-walk — APPROVED** [review]

**Comment by bog-walk** [comment]
From my understanding, PostgreSQL will be the only database that returns a result set populated with inserted/updated values that are not auto-incrementing. It might be good to investigate later if we can instead return a null `resultValues` for those databases, instead of an incorrect UUID. But I'm not sure how feasible that is as it would mean needing to determine whether the upsert was successful as an insert or an update (which I believe is only possible with MySQL), or risk breaking the expected behavior of inserts having their result sets overriden.

**Comment by e5l** [comment]
@bog-walk, if we're returning invalid UUID for some DBs it should be fixed before merging

**Review by e5l — APPROVED** [review]

**Pull request merged**